src/sql_generator.cpp

```diff
@@ -219,48 +298,90 @@ void alter_table(duckdb::Connection &con, const table_def &table,
-    auto result = con.Query(query);
-    if (result->HasError()) {
-      throw std::runtime_error("Could not drop column <" + col_name +
-                               "> from table <" + absolute_table_name +
-                               ">:" + result->GetError());
-    }
+    run_query(con, "alter_table drop", out.str(),
+              "Could not drop column <" + col_name + "> from table <" +
+                  absolute_table_name + ">");
   }
 
   for (const auto &col_name : alter_types) {
     std::ostringstream out;
     out << "ALTER TABLE " << absolute_table_name << " ALTER ";
-    const auto &col = new_column_map[col_name];
+    const auto &col = new_column_map.at(col_name);
 
     out << KeywordHelper::WriteQuoted(col_name, '"') << " TYPE "
         << duckdb::EnumUtil::ToChars(col.type);
 
-    auto query = out.str();
-    mdlog::info("alter table: " + query);
-    auto result = con.Query(query);
-    if (result->HasError()) {
-      throw std::runtime_error("Could not alter type for column <" + col_name +
-                               "> in table <" + absolute_table_name +
-                               ">:" + result->GetError());
-    }
+    run_query(con, "alter table change type", out.str(),
+              "Could not alter type for column <" + col_name + "> in table <" +
+                  absolute_table_name + ">");
   }
 }
 
-std::string
-make_full_column_list(const std::vector<const column_def *> &columns_pk,
-                      const std::vector<const column_def *> &columns_regular) {
-  std::ostringstream full_column_list;
-  if (!columns_pk.empty()) {
-    write_joined(full_column_list, columns_pk, print_column);
-    // tiny troubleshooting assist; primary columns are separated from regular
-    // columns by 2 spaces
-    full_column_list << ",  ";
+void alter_table(duckdb::Connection &con, const table_def &table,
+                 const std::vector<column_def> &columns) {
+
+  bool recreate_table = false;
+
+  auto absolute_table_name = table.to_escaped_string();
+  std::set<std::string> alter_types;
+  std::set<std::string> added_columns;
+  std::set<std::string> deleted_columns;
+  std::set<std::string> common_columns;
+
+  const auto &existing_columns = describe_table(con, table);
+  std::map<std::string, column_def> new_column_map;
+
+  for (const auto &col : columns) {
+    new_column_map.emplace(col.name, col);
+    added_columns.emplace(col.name);
   }
-  write_joined(full_column_list, columns_regular, print_column);
 
-  return full_column_list.str();
+  for (const auto &col : existing_columns) {
+    const auto &new_col_it = new_column_map.find(col.name);
+
+    if (added_columns.erase(col.name)) {
+      common_columns.emplace(col.name);
+    }
+
+    if (new_col_it == new_column_map.end()) {
+      deleted_columns.emplace(col.name);
+      if (col.primary_key) {
+        recreate_table = true;
+      }
+    } else if (new_col_it->second.type != col.type) {
+      alter_types.emplace(col.name);
```

Comment:
why can a change of type happen in place?

Reply:
or rather, why does this check come before the `new_col_it->second.primary_key != col.primary_key` check, and take precedence over change of primary key status

Reply:
if the types are compatible, it will work. If they are not, this should not be happening in the first place, and recreating a table won't really help
```
D create table a (col short);
D insert into a (col) values (1), (2), (3);
D select * from a;
┌───────┐
│  col  │
│ int16 │
├───────┤
│     1 │
│     2 │
│     3 │
└───────┘
D alter table a alter col type INT;
D select * from a;
┌───────┐
│  col  │
│ int32 │
├───────┤
│     1 │
│     2 │
│     3 │
└───────┘
```

Reply:
oh, that's a good point; I'll switch them

Reply:
done!